### PR TITLE
Move detection of self mapping IDs to IterDomainGraph from ComputeAtRootDomainMap

### PR DESCRIPTION
**Pull request by naoyam**
ComputeAtRootDomainMap is meant to detect domains that are unsafe to map
with computeAt. One scenario of unsafe mapping is the reduction and
broadcast patten commonly appearing in normalizations. Another scenario
is when a domain is mapped with multiple domains of another tensor,
which means that all of the domains  must be mapped together, i.e.,
grouped into the same for-loop, but grouping multiple domains of the
same tensor into the same for-loop doesn't seem right.

Previously, ComputeAtRootDomainMap tried to detect the above pattern of mappings
by a backward domain analysis, which has been working fine, but with
views, it resulted in false positives due to an approximated analysis
for the detection of the reduction-broadcast pattern. This happened with
the FusionViewMagicSchedule3 test in #1928.

In this PR, since self-mapping domains are more generic problems, not
specific to computeAt, I moved the validation to IterDomainGraph,
specifically, I added the validation function to the end of the graph
builder. I feel that the validation may be a little too much leaned
towards the safe side as it may not be necessary to look at both root
and leaf domains, but I'm not sure if we could just look at root or leaf
domains.

I also updated some of the relevant tests accordingly. We still do not
support fusions like FusionDetectSelfMappings. The first failure I saw
happened in indexing, but I suspect that's not the only place we have
this assumption.

Moving forward, it may make more sense to move the remaining logic of
ComputeAtRootDomainMap to work on IterDomainGraph, but I do not see any
immediate priority to do so as of now.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/disjoint_set.h [hunk 1]
```diff
@@ -302,17 +302,14 @@ class DisjointSets {
...
       for (auto entry : set.vector()) {
-        ss << abstractToString(entry);
-        // DomainKey defines == but not !=
-        if (!(entry == set.back())) {
-          ss << "; ";
-        }
+        ss << sep << sep << abstractToString(entry) << "\n";
       }
-      ss << " }\n";
+      ss << sep << "}\n";
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/disjoint_set.h:312]: Just make it look better when printed

File: torch/csrc/jit/codegen/cuda/root_domain_map.cpp [hunk 2]
```diff
@@ -186,27 +186,39 @@ auto ensureMapping(
...
+  ss << " in ";
+  if (td()) {
+    auto tv = lookUpTv(td());
+    TORCH_INTERNAL_ASSERT(tv != nullptr, "No TV found for ", td()->toString());
+    ss << "T" << tv->name() << "[ " << td()->getRootDomain() << " ]";
+    if (td()->hasRFactor()) {
+      ss << " (Rfactor: [ " << td()->getMaybeRFactorDomain() << " ])";
+    }
+  } else {
+    ss << "null";
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/root_domain_map.cpp:220]: Ditto. Making it look better

File: torch/csrc/jit/codegen/cuda/test/test_gpu.cpp [hunk 3]
```diff
@@ -10867,26 +10838,6 @@ TEST_F(NVFuserTest, FusionLSTMCell_CUDA) {
...
-  TensorView* tv3 = broadcast(tv1, {false, true});
-  TensorView* tv4 = add(tv2, tv3);
-  fusion.addOutput(tv4);
-
-  // Not possible to do computeAt at position -1 as recomputation
-  // would be required. An exception should be thrown.
-  // NOLINTNEXTLINE(cppcoreguidelines-avoid-goto,hicpp-avoid-goto)
-  ASSERT_ANY_THROW(tv1->computeAt(tv3, -1));
-}
-
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/test/test_gpu.cpp:10889]: This exercises almost the same pattern as FusionDetectSelfMappedDomains, so removed

**Review by zasdfgbnm — COMMENTED** [review]
@naoyam Let's move the discussion here. Does it make sense not to throw a hard error for the self-mapping case, but instead, make it a method `hasSelfMapping`? Doing so allows us to modify this function
https://github.com/csarofeen/pytorch/blob/631094891a96f715d8c9925fb73d41013ca7f2e3/torch/csrc/jit/codegen/cuda/scheduler/registry.cpp#L1304-L1325
to add a check
```c++
if (id_graph.hasSelfMapping()) {
  return false;
}
```
And this will cause the segmenter to segment the fusion into pieces that do not have self-mapping. And we don't need to do anything in the lowering to get "self-mapping support" for free.

It don't have to be in this PR though.

**Comment by naoyam** [comment]
> @naoyam Let's move the discussion here. Does it make sense not to throw a hard error for the self-mapping case, but instead, make it a method `hasSelfMapping`? Doing so allows us to modify this function
> 
> https://github.com/csarofeen/pytorch/blob/631094891a96f715d8c9925fb73d41013ca7f2e3/torch/csrc/jit/codegen/cuda/scheduler/registry.cpp#L1304-L1325
> 
> 
> to add a check
> ```c++
> if (id_graph.hasSelfMapping()) {
>   return false;
> }
> ```
> 
> And this will cause the segmenter to segment the fusion into pieces that do not have self-mapping. And we don't need to do anything in the lowering to get "self-mapping support" for free.
> 
> It don't have to be in this PR though.

I thought the default behavior or `IterDomainGraph` should always error out if self mapping is detected, but maybe it should be relaxed, and checking should be opt-in? I think `IterDomainGraph` is used in many places, so we would need to examine each of them if the check should be done or not.

**Comment by csarofeen** [comment]
I'm nervous to support any self mapping of any concrete iter domains. What do we want self mapping support for?

**Review by csarofeen — APPROVED** [review]
LGTM, merging as is, so I can merge the change into view scheduling.

**Pull request merged**